### ISSUE TITLE
(File.Error) could not write to file ".git/info/sparse-checkout": no such file or directory

### ISSUE THREAD

**Issue by albertoalmagro** [elixir-lang/elixir#5525]
### Environment

* Elixir version: Elixir 1.3.4 (Local Elixir)
* Erlang version: Erlang/OTP 19 [erts-8.1] [source] [64-bit] [smp:4:4] [async-threads:10] [hipe] [kernel-poll:false] [dtrace]
* Tried to build source code with last commit: f6a5d80b578978b940d10b0a56c1e6088a34ed85
* Operating system: Mac OS X El Capitan 10.11.6
* Git version: 2.9.0

### Current behavior

When cloning and trying to build Elixir's source code 3 tests fail with (File.Error) could not write to file ".git/info/sparse-checkout": no such file or directory.

Executed
```
git clone https://github.com/elixir-lang/elixir.git
cd elixir
make clean test
```
The following error is thrown:
```
  1) test updates the repo when sparse is turned on (Mix.Tasks.DepsGitTest)
     test/mix/tasks/deps.git_test.exs:313
     ** (File.Error) could not write to file ".git/info/sparse-checkout": no such file or directory
     stacktrace:
       (elixir) lib/file.ex:719: File.write!/3
       (mix) lib/mix/scm/git.ex:104: Mix.SCM.Git.checkout/2
       (elixir) lib/file.ex:1162: File.cd!/2
       (mix) lib/mix/dep/fetcher.ex:61: Mix.Dep.Fetcher.do_fetch/3
       (mix) lib/mix/dep/converger.ex:177: Mix.Dep.Converger.all/9
       (mix) lib/mix/dep/converger.ex:116: Mix.Dep.Converger.all/7
       (mix) lib/mix/dep/converger.ex:71: Mix.Dep.Converger.all/4
       (mix) lib/mix/dep/converger.ex:50: Mix.Dep.Converger.converge/4
       (mix) lib/mix/dep/fetcher.ex:16: Mix.Dep.Fetcher.all/3
       (mix) lib/mix/tasks/deps.get.ex:27: Mix.Tasks.Deps.Get.run/1
       test/mix/tasks/deps.git_test.exs:328: anonymous fn/0 in Mix.Tasks.DepsGitTest.test updates the repo when sparse is turned on/1
       (elixir) lib/file.ex:1162: File.cd!/2
       test/test_helper.exs:108: MixTest.Case.in_fixture/3
       test/mix/tasks/deps.git_test.exs:316: (test)

.........

  2) test gets and updates Git repos with sparse checkout (Mix.Tasks.DepsGitTest)
     test/mix/tasks/deps.git_test.exs:73
     ** (File.Error) could not write to file ".git/info/sparse-checkout": no such file or directory
     stacktrace:
       (elixir) lib/file.ex:719: File.write!/3
       (mix) lib/mix/scm/git.ex:104: Mix.SCM.Git.checkout/2
       (elixir) lib/file.ex:1162: File.cd!/2
       (mix) lib/mix/dep/fetcher.ex:64: Mix.Dep.Fetcher.do_fetch/3
       (mix) lib/mix/dep/converger.ex:177: Mix.Dep.Converger.all/9
       (mix) lib/mix/dep/converger.ex:116: Mix.Dep.Converger.all/7
       (mix) lib/mix/dep/converger.ex:71: Mix.Dep.Converger.all/4
       (mix) lib/mix/dep/converger.ex:50: Mix.Dep.Converger.converge/4
       (mix) lib/mix/dep/fetcher.ex:16: Mix.Dep.Fetcher.all/3
       (mix) lib/mix/tasks/deps.get.ex:27: Mix.Tasks.Deps.Get.run/1
       test/mix/tasks/deps.git_test.exs:78: anonymous fn/0 in Mix.Tasks.DepsGitTest.test gets and updates Git repos with sparse checkout/1
       (elixir) lib/file.ex:1162: File.cd!/2
       test/test_helper.exs:108: MixTest.Case.in_fixture/3
       test/mix/tasks/deps.git_test.exs:77: (test)

..

  3) test updates the repo when sparse is turned off (Mix.Tasks.DepsGitTest)
     test/mix/tasks/deps.git_test.exs:284
     ** (File.Error) could not write to file ".git/info/sparse-checkout": no such file or directory
     stacktrace:
       (elixir) lib/file.ex:719: File.write!/3
       (mix) lib/mix/scm/git.ex:104: Mix.SCM.Git.checkout/2
       (elixir) lib/file.ex:1162: File.cd!/2
       (mix) lib/mix/dep/fetcher.ex:64: Mix.Dep.Fetcher.do_fetch/3
       (mix) lib/mix/dep/converger.ex:177: Mix.Dep.Converger.all/9
       (mix) lib/mix/dep/converger.ex:116: Mix.Dep.Converger.all/7
       (mix) lib/mix/dep/converger.ex:71: Mix.Dep.Converger.all/4
       (mix) lib/mix/dep/converger.ex:50: Mix.Dep.Converger.converge/4
       (mix) lib/mix/dep/fetcher.ex:16: Mix.Dep.Fetcher.all/3
       (mix) lib/mix/tasks/deps.get.ex:27: Mix.Tasks.Deps.Get.run/1
       test/mix/tasks/deps.git_test.exs:289: anonymous fn/0 in Mix.Tasks.DepsGitTest.test updates the repo when sparse is turned off/1
       (elixir) lib/file.ex:1162: File.cd!/2
       test/test_helper.exs:108: MixTest.Case.in_fixture/3
       test/mix/tasks/deps.git_test.exs:288: (test)

................................................................................................................................................................................................................

Finished in 94.3 seconds (6.8s on load, 87.5s on tests)
401 tests, 3 failures

```

After that I checked Erlang version and tried to build again with commands:
```
make clean
make test
```
But same result.

Finally I tried to compile the code and test again with commands:
```
make clean compile
make test
```
But same result.

Directory does .git/info does not exist, but maybe it should be created in that case.
```
alberto.almagro@wsp0169 elixir (master) $ ls -la .git
total 168
drwxr-xr-x  12 alberto.almagro  679754705    408  3 dic 23:27 .
drwxr-xr-x  22 alberto.almagro  679754705    748  3 dic 16:58 ..
-rw-r--r--   1 alberto.almagro  679754705    649  3 dic 23:27 FETCH_HEAD
-rw-r--r--   1 alberto.almagro  679754705     23  3 dic 16:58 HEAD
-rw-r--r--   1 alberto.almagro  679754705     41  3 dic 23:27 ORIG_HEAD
-rw-r--r--   1 alberto.almagro  679754705    311  3 dic 16:58 config
drwxr-xr-x   3 alberto.almagro  679754705    102  3 dic 16:58 hooks
-rw-r--r--   1 alberto.almagro  679754705  60220  3 dic 23:27 index
drwxr-xr-x   4 alberto.almagro  679754705    136  3 dic 16:58 logs
drwxr-xr-x  53 alberto.almagro  679754705   1802  3 dic 23:27 objects
-rw-r--r--   1 alberto.almagro  679754705   4210  3 dic 16:58 packed-refs
drwxr-xr-x   5 alberto.almagro  679754705    170  3 dic 16:58 refs
```

### Expected behavior

Elixir source code is successfully built.

If this bug is confirmed I would like to write a PR to fix it. Thanks.

**Comment by josevalim** [comment]
The tests do not fail locally nor in the CI server, so if you can debug why
it happens on your machine, it would be appreciated. I have Git 2.9.3.

**Comment by albertoalmagro** [comment]
Thanks José. It was happening on my local machine.

Updating my local git version solved the issue, now with git version 2.11.0 it doesn't happen anymore. We can close this issue if you want.

Now I am getting another error on test/elixir/system_test.exs:16 f6a5d80b578978b940d10b0a56c1e6088a34ed85 But that would be another ticket or local issue.

When you told me to debug the application, I've tried to do it with IEx.pry, has I usually do on my apps, but it didn't stop for the source code. Is it another way to debug it? Thanks in advance.

**Comment by josevalim** [comment]
You should be able to use IEx.pry as usual in Elixir too. Did you get an error or something in particular?

**Comment by albertoalmagro** [comment]
Thank José. At `elixir/lib/mix/test/mix/tasks/deps.git_test.exs` I first required IEx

```
Code.require_file "../../test_helper.exs", __DIR__
require IEx;

defmodule Mix.Tasks.DepsGitTest do
  use MixTest.Case
```

And then called it within the test code:

 ```
 @tag :git_sparse
  test "gets and updates Git repos with sparse checkout" do
    Process.put(:git_repo_opts, sparse: "sparse_dir")
    Mix.Project.push GitApp
    IEx.pry

    in_fixture "no_mixfile", fn ->
      IEx.pry
      Mix.Tasks.Deps.Get.run []
      message = "* Getting git_repo (#{fixture_path("git_repo")})"
      assert_received {:mix_shell, :info, [^message]}
      refute File.exists?("deps/git_repo/mix.exs")
      assert File.exists?("deps/git_repo/sparse_dir/mix.exs")
      assert File.read!("mix.lock") =~ "sparse: \"sparse_dir\""
    end
  end
```
After that I ran again `make clean test` but it doesn't stop.

**Comment by josevalim** [comment]
You don't need to run `make clean test`. You should be able to run it as `iex lib/mix/test/mix/tasks/deps.git_test.exs` from the Elixir root.

**Comment by bryanstearns** [comment]
TLDR: if you have a global git repository template, make sure it has an `info` subdirectory.

For others who find this issue: I had this problem today, and I'm already running git 2.11.0 ... I tracked it down to my having a global `~/.gitconfig` that pointed at my git template (which has a hook script that complains if I try to check in trailing whitespace, etc)

```
[init]
  templatedir = ~/.git_template
```
It turns out that if you specify a global init template like this, `git init` doesn't create certain things in the new repository:
- a `hooks` subdirectory with a bunch of disabled sample hook scripts
- a `description` file containing "Unnamed repository: edit this file 'description' to name the respository."
- an `info` subdirectory containing `exclude`, with several commented out example exclusions.

Of course, the `mix` tests use `git init`, and assume that the `info` subdirectory will be created by `git init`, causing these failures.

I created an empty `info` subdirectory  in my template and the tests all pass again. I don't think it's worth fixing the tests to deal with this, but it'd be an easy change to mix's [test_helper.exs](https://github.com/elixir-lang/elixir/blob/90f368802f1cfbc22398dff3ecb368c325b77cb4/lib/mix/test/test_helper.exs#L207) to `mkdir_p` the info directory after `git init` -- let me know if you'd like a PR.

**Comment by josevalim** [comment]
Thank you for reporting back @bryanstearns. I am afraid that using a mkdir_p is not enough as anything people put in their templates can affect `git init`. Is there a way to call `git init` and tell it to ignore any existing template?

**Comment by bryanstearns** [comment]
Ooh, good thinking, @josevalim -- there doesn't seem to be a way to say "no template", but several possible ways of overriding the global setting. I'll experiment and report back.

**Comment by bryanstearns** [comment]
@josevalim ... Nope, git doesn't provide a way to say "no template". Instead, I spent a bit of time trying to provide our own dummy template to override the user's, but just realized that this needs to happen in Mix.SCM.Git, not in `test_helper.exs`, so I think that's a non-starter.

Instead, we could add `File.mkdir_p!(".git/info")` before [this line in `Mix.SCM.Git.sparse_toggle/1`](https://github.com/elixir-lang/elixir/blob/90f368802f1cfbc22398dff3ecb368c325b77cb4/lib/mix/lib/mix/scm/git.ex#L142). That code is already coupled to git's repo structure, and we'd just making it more resilient. (I'm open to suggestions on how to write a test for this change, though!)

Or, of course, we could do nothing: this issue is the first hit if you Google the error message.

**Comment by josevalim** [comment]
@bryanstearns please send a PR where we use mkdir_p! before writing to sparse checkout. Given we are writing to that directory, I believe it makes sense to ensure it exists.

**Comment by albertoalmagro** [comment]
@bryanstearns Thanks for the fix!
